File: app_watcher/src/main/java/com/tongchen/twatcher/MainActivity.java
package com.tongchen.twatcher;


import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.tongchen.twatcher.base.ui.BackHandleInterface;
import com.tongchen.twatcher.base.ui.activity.BaseActivity;
import com.tongchen.twatcher.base.ui.fragment.BaseFragment;
import com.tongchen.twatcher.gank.ui.fragment.GankMainFragment;
import com.tongchen.twatcher.mzitu.ui.fragment.MZiTuFragment;
import com.tongchen.twatcher.util.LogUtils;
import com.tongchen.twatcher.util.ToastUtils;
import com.tongchen.twatcher.widget.TDrawerLayout;

import butterknife.BindView;
import butterknife.OnClick;

public class MainActivity extends BaseActivity implements BackHandleInterface {

    private static final String TAG = "MainActivity";

    //  不同分类的最底层Fragment
    public static final String TAG_FRAGMENT_MAIN = "MainFragment";

    @BindView(R.id.tdrawerlyt)
    TDrawerLayout mTDrawerLyt;
    @BindView(R.id.ll_menu_container)
    LinearLayout mMenuContainerLLyt;
    @BindView(R.id.tv_gank)
    TextView mGankTv;
    @BindView(R.id.tv_mzitu)
    TextView mMZiTuTv;
    @BindView(R.id.fl_content_container)
    FrameLayout mContentContainerFLyt;

    private FragmentManager mFragmentManager;
    private GankMainFragment mGankFragment;
    private MZiTuFragment mMZiTuFragment;

    //  点击2次返回才退出
    private long firstTime = 0;

    private BaseFragment mBaseFragment;

    @Override
    public int bindLayout() {
        return R.layout.main_activity;
    }

    @Override
    protected void loadView() {
        mFragmentManager = getSupportFragmentManager();
        mGankFragment = GankMainFragment.newInstance();
        mMZiTuFragment = MZiTuFragment.newInstance();

        loadFragment(mGankFragment);

    }

    private void loadFragment(Fragment fragment) {
        FragmentTransaction transaction = mFragmentManager.beginTransaction();
        transaction.add(R.id.fl_content_container, fragment, TAG_FRAGMENT_MAIN);
        transaction.commit();
    }

    @OnClick({R.id.tv_gank, R.id.tv_mzitu})
    public void onViewClick(View view) {
        LogUtils.d(TAG, "view---" + view.toString());
        switch (view.getId()) {
            case R.id.tv_gank:
                if (mGankFragment == null) {
                    mGankFragment = GankMainFragment.newInstance();
                }
                if (!mGankFragment.isAdded()) {
                    loadFragment(mGankFragment);
                    mFragmentManager.beginTransaction().hide(mMZiTuFragment).commit();
                } else {
                    if (!mGankFragment.isVisible()) {
                        mFragmentManager.beginTransaction().hide(mMZiTuFragment).show(mGankFragment).commit();
                    }
                }

                break;
            case R.id.tv_mzitu:
                if (mMZiTuFragment == null) {
                    mMZiTuFragment = MZiTuFragment.newInstance();
                }
                if (!mMZiTuFragment.isAdded()) {
                    loadFragment(mMZiTuFragment);
                    mFragmentManager.beginTransaction().hide(mGankFragment).commit();
                } else {
                    if (!mMZiTuFragment.isVisible()) {
                        mFragmentManager.beginTransaction().hide(mGankFragment).show(mMZiTuFragment).commit();
                    }
                }
                break;
        }
    }

    @Override
    public void onBackPressed() {
        if (mBaseFragment == null || !mBaseFragment.onBackPressed()) {
            if (getSupportFragmentManager().getBackStackEntryCount() == 0) {
                //  如果当前菜单展开，先隐藏菜单
                if (mTDrawerLyt.isMenuOpen()) {
                    mTDrawerLyt.toggleMenu();
                    return;
                }
                long secondTime = System.currentTimeMillis();
                if (secondTime - firstTime < 2000) {
                    System.exit(0);
                } else {
                    ToastUtils.showShort(R.string.sys_exit);
                    firstTime = System.currentTimeMillis();
                }
            } else {
                getSupportFragmentManager().popBackStack();
            }
        }

    }

    public void isDispatchEvent2DrawerLayout(boolean isDispatch) {
        mTDrawerLyt.setConsume(isDispatch);
    }

    public void toggleDrawerLyt() {
        mTDrawerLyt.toggleMenu();
    }

    @Override
    public void setSelectedFragment(BaseFragment selectedFragment) {
        mBaseFragment = selectedFragment;
    }

    public void startFragment(Fragment fragment) {
//        mTDrawerLyt.setConsume(false);
        FragmentTransaction transaction = mFragmentManager.beginTransaction();
        transaction.setCustomAnimations(R.anim.fragment_translate_in, R.anim.fragment_translate_out, R.anim.fragment_translate_in, R.anim.fragment_translate_out);
        transaction.add(R.id.fl_content_container, fragment).addToBackStack(null);
        transaction.commit();
    }


}

File: app_watcher/src/main/java/com/tongchen/twatcher/MainActivity.kt
package com.tongchen.twatcher

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.view.View
import com.tongchen.twatcher.base.ui.BackHandleInterface
import com.tongchen.twatcher.base.ui.activity.BaseActivity
import com.tongchen.twatcher.base.ui.fragment.BaseFragment
import com.tongchen.twatcher.gank.ui.fragment.GankMainFragment
import com.tongchen.twatcher.mzitu.ui.fragment.MZiTuMainFragment
import com.tongchen.twatcher.util.ToastUtils
import kotlinx.android.synthetic.main.main_activity.*
import kotlinx.android.synthetic.main.main_include_menu.*
import kotlin.system.exitProcess

/**
 * @author TongChen
 * @date 2019/11/29  18:30
 * <p>
 * Desc:
 */
class MainActivity : BaseActivity(), BackHandleInterface, View.OnClickListener {

    private val TAG = "MainActivity"
    //  不同分类的最底层Fragment
    private val TAG_FRAGMENT_MAIN = "MainFragment"

    private lateinit var mFragmentManager: FragmentManager
    private var mGankMainFragment: GankMainFragment? = null
    private var mMZiTuMainFragment: MZiTuMainFragment? = null
    private var mBaseFragment: BaseFragment? = null
    //  点击2次返回才退出
    private var firstTime: Long = 0

    override fun bindLayout() = R.layout.main_activity

    override fun loadView() {
        tv_gank.setOnClickListener(this)
        tv_mzitu.setOnClickListener(this)

        mFragmentManager = supportFragmentManager
        mGankMainFragment = GankMainFragment.newInstance()
        mMZiTuMainFragment = MZiTuMainFragment.newInstance()

        loadFragment(mGankMainFragment as Fragment)
    }

    private fun loadFragment(fragment: Fragment) {
        val transaction = mFragmentManager.beginTransaction()
        transaction.add(R.id.fl_content_container, fragment, TAG_FRAGMENT_MAIN)
        transaction.commit()
    }

    override fun onClick(v: View?) {
        when (v?.id) {
            R.id.tv_gank -> {
                if (mGankMainFragment == null) {
                    mGankMainFragment = GankMainFragment.newInstance()
                }
                if (!mGankMainFragment!!.isAdded) {
                    loadFragment(mGankMainFragment as Fragment)
                    mFragmentManager.beginTransaction().hide(mMZiTuMainFragment as Fragment).commit()

                } else {
                    if (!mGankMainFragment!!.isVisible) {
                        mFragmentManager.beginTransaction().hide(mMZiTuMainFragment as Fragment).show(mGankMainFragment as Fragment).commit()
                    }
                }
            }

            R.id.tv_mzitu -> {
                if (mMZiTuMainFragment == null) {
                    mMZiTuMainFragment = MZiTuMainFragment.newInstance()
                }
                if (!mMZiTuMainFragment!!.isAdded) {
                    loadFragment(mMZiTuMainFragment as Fragment)
                    mFragmentManager.beginTransaction().hide(mGankMainFragment as Fragment).commit()

                } else {
                    if (!mMZiTuMainFragment!!.isVisible) {
                        mFragmentManager.beginTransaction().hide(mGankMainFragment as Fragment).show(mMZiTuMainFragment as Fragment).commit()
                    }
                }
            }
        }

        toggleDrawerLyt()
    }

    override fun onBackPressed() {
        if (mBaseFragment == null || !mBaseFragment!!.onBackPressed()) {
            if (supportFragmentManager.backStackEntryCount == 0) {
                //  如果当前菜单展开，先隐藏菜单
                if (tdrawerlyt.isMenuOpen) {
                    tdrawerlyt.toggleMenu()
                    return
                }
                val secondTime = System.currentTimeMillis()
                if (secondTime - firstTime < 2000) {
                    exitProcess(0)
                } else {
                    ToastUtils.showShort(R.string.sys_exit)
                    firstTime = System.currentTimeMillis()
                }
            } else {
                supportFragmentManager.popBackStack()
            }
        }
    }

    fun isDispatchEvent2DrawerLayout(isDispatch: Boolean) {
        tdrawerlyt.setConsume(isDispatch)
    }

    fun toggleDrawerLyt() {
        tdrawerlyt.toggleMenu()
    }

    override fun setSelectedFragment(selectedFragment: BaseFragment?) {
        mBaseFragment = selectedFragment
    }

    fun startFragment(fragment: Fragment) {
        //  mTDrawerLyt.setConsume(false);
        val transaction = mFragmentManager.beginTransaction()
        transaction.setCustomAnimations(R.anim.fragment_translate_in, R.anim.fragment_translate_out, R.anim.fragment_translate_in, R.anim.fragment_translate_out)
        transaction.add(R.id.fl_content_container, fragment).addToBackStack(null)
        transaction.commit()
    }
}